File: scripts/translate_to_arabic.py
import json
import time
import sys
from googletrans import Translator

TRANSLATIONS_PATH = 'frontend/assets/translations.json'

with open(TRANSLATIONS_PATH, 'r', encoding='utf-8') as f:
    translations = json.load(f)

SKIP_PATTERNS = [
    r'^[\d\s\-\+\$\.\,\(\)\%]+$',
    r'^\[email&#160;protected\]$',
    r'^©?\s*Pentu\.?\s*$',
    r'^Pentu\s*-',
    r'^\d{1,2}\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d{4}$',
    r'^\d+\s+(Comment|Comments)$',
    r'^\$[\d\.,]+$',
    r'^[A-Z][a-z]+\s+[A-Z][a-z]+$',
    r'^[A-Z][a-z]+\s+[A-Z][a-z]+\s+&\s+[A-Z][a-z]+$',
    r'^info@[a-z\.]+$',
]

import re

def should_skip(text):
    if not text or len(text) < 2:
        return True
    for pattern in SKIP_PATTERNS:
        if re.match(pattern, text):
            return True
    return False

shared_keys = ['site', 'header', 'nav', 'hero', 'services', 'about', 'footer', 'common']
unique_texts = {}

for key, en_val in translations['en'].items():
    ar_val = translations['ar'].get(key)
    is_shared = any(key == p or key.startswith(p + '.') for p in shared_keys)
    if is_shared and en_val != ar_val:
        continue
    if en_val == ar_val and not should_skip(en_val):
        if en_val not in unique_texts:
            unique_texts[en_val] = []
        unique_texts[en_val].append(key)

texts_to_translate = list(unique_texts.keys())
print(f'Found {len(texts_to_translate)} unique strings to translate')

if len(texts_to_translate) == 0:
    print('Nothing to translate!')
    sys.exit(0)

translator = Translator()
translated = {}
BATCH_SIZE = 20

for i in range(0, len(texts_to_translate), BATCH_SIZE):
    batch = texts_to_translate[i:i+BATCH_SIZE]
    print(f'Translating batch {i//BATCH_SIZE + 1}/{(len(texts_to_translate) + BATCH_SIZE - 1)//BATCH_SIZE} ({len(batch)} strings)...')
    
    for text in batch:
        try:
            result = translator.translate(text, src='en', dest='ar')
            translated[text] = result.text
            print(f'  OK: {text[:50]} -> {result.text[:50]}')
        except Exception as e:
            print(f'  ERROR: {text[:50]} -> {e}', file=sys.stderr)
            translated[text] = text
        time.sleep(0.5)
    
    if i + BATCH_SIZE < len(texts_to_translate):
        print('  Waiting 3 seconds...')
        time.sleep(3)

updated_count = 0
for text, keys in unique_texts.items():
    ar_text = translated.get(text, text)
    for key in keys:
        translations['ar'][key] = ar_text
        updated_count += 1

def deep_sort(obj):
    if not isinstance(obj, dict):
        return obj
    return {k: deep_sort(v) for k, v in sorted(obj.items())}

translations['en'] = deep_sort(translations['en'])
translations['ar'] = deep_sort(translations['ar'])

with open(TRANSLATIONS_PATH, 'w', encoding='utf-8') as f:
    json.dump(translations, f, ensure_ascii=False, indent=2)

print(f'\n✓ Updated {updated_count} translation keys with Arabic text')
print(f'✓ Saved to translations.json')

File: scripts/translate-to-arabic.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const TRANSLATIONS_PATH = path.join(__dirname, '..', 'frontend', 'assets', 'translations.json');
const translations = JSON.parse(fs.readFileSync(TRANSLATIONS_PATH, 'utf8'));

// Strings to skip from auto-translation (proper names, template artifacts, code)
const SKIP_PATTERNS = [
  /^[\d\s\-\+\$\.\,\(\)\%]+$/, // pure numbers/money
  /^\[email&#160;protected\]$/, // email protection artifact
  /^©?\s*Pentu\.?\s*$/, // old template brand
  /^Pentu\s*-/, // old template title
  /^\d{1,2}\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d{4}$/, // dates
  /^\d+\s+(Comment|Comments)$/, // comment counts
  /^\$[\d\.,]+$/, // prices
  /^[A-Z][a-z]+\s+[A-Z][a-z]+$/, // proper names (John Smith)
  /^[A-Z][a-z]+\s+[A-Z][a-z]+\s+&\s+[A-Z][a-z]+$/, // names with &
  /^info@[a-z\.]+$/, // emails
];

function shouldSkip(text) {
  if (!text || text.length < 2) return true;
  for (const pattern of SKIP_PATTERNS) {
    if (pattern.test(text)) return true;
  }
  return false;
}

// Collect unique English strings that need translation
const uniqueTexts = new Map(); // text -> [keys]
const sharedKeys = ['site', 'header', 'nav', 'hero', 'services', 'about', 'footer', 'common'];

for (const [key, enVal] of Object.entries(translations.en)) {
  const arVal = translations.ar[key];
  // Skip already-translated shared keys
  const isShared = sharedKeys.some(p => key === p || key.startsWith(p + '.'));
  if (isShared && enVal !== arVal) continue;
  
  if (enVal === arVal && !shouldSkip(enVal)) {
    if (!uniqueTexts.has(enVal)) uniqueTexts.set(enVal, []);
    uniqueTexts.get(enVal).push(key);
  }
}

const textsToTranslate = Array.from(uniqueTexts.keys());
console.log(`Found ${textsToTranslate.length} unique strings to translate`);

if (textsToTranslate.length === 0) {
  console.log('Nothing to translate!');
  process.exit(0);
}

// Use Python googletrans for translation
function translateBatch(texts) {
  const pythonScript = `
from googletrans import Translator
import json
import sys

translator = Translator()
texts = json.loads(sys.argv[1])
results = {}
for text in texts:
    try:
        result = translator.translate(text, src='en', dest='ar')
        results[text] = result.text
    except Exception as e:
        results[text] = text  # fallback
        print(f"ERROR: {text} -> {e}", file=sys.stderr)
print(json.dumps(results, ensure_ascii=False))
`;
  const inputJson = JSON.stringify(texts);
  try {
    const result = execSync(`python -c "${pythonScript.replace(/"/g, '\\"').replace(/\n/g, '; ')}" "${inputJson.replace(/"/g, '\\"')}"`, {
      encoding: 'utf8',
      timeout: 120000,
      maxBuffer: 10 * 1024 * 1024
    });
    return JSON.parse(result);
  } catch (e) {
    console.error('Translation batch failed:', e.stderr || e.message);
    const fallback = {};
    for (const t of texts) fallback[t] = t;
    return fallback;
  }
}

// Translate in batches of 30
const BATCH_SIZE = 30;
const translated = {};

for (let i = 0; i < textsToTranslate.length; i += BATCH_SIZE) {
  const batch = textsToTranslate.slice(i, i + BATCH_SIZE);
  console.log(`Translating batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(textsToTranslate.length / BATCH_SIZE)} (${batch.length} strings)...`);
  
  const batchResults = translateBatch(batch);
  Object.assign(translated, batchResults);
  
  // Small delay between batches
  if (i + BATCH_SIZE < textsToTranslate.length) {
    console.log('  Waiting 2 seconds...');
    execSync('node -e "setTimeout(() => process.exit(0), 2000)"');
  }
}

// Apply translations
let updatedCount = 0;
for (const [text, keys] of uniqueTexts) {
  const arText = translated[text] || text;
  for (const key of keys) {
    translations.ar[key] = arText;
    updatedCount++;
  }
}

// Sort and save
function deepSort(obj) {
  if (typeof obj !== 'object' || obj === null) return obj;
  if (Array.isArray(obj)) return obj.map(deepSort);
  const sorted = {};
  for (const key of Object.keys(obj).sort()) {
    sorted[key] = deepSort(obj[key]);
  }
  return sorted;
}

translations.en = deepSort(translations.en);
translations.ar = deepSort(translations.ar);

fs.writeFileSync(TRANSLATIONS_PATH, JSON.stringify(translations, null, 2), 'utf8');
console.log(`\n✓ Updated ${updatedCount} translation keys with Arabic text`);
console.log(`✓ Saved to translations.json`);
